src/coreclr/jit/lclmasks.cpp

```diff
@@ -0,0 +1,480 @@
+//
+// Arguments:
+//     op - The HW intrinsic to cache
+//
+void LclMasksWeight::CacheSimdTypes(GenTreeHWIntrinsic* op)
+{
+    CorInfoType newSimdBaseJitType = op->GetSimdBaseJitType();
+    unsigned    newSimdSize        = op->GetSimdSize();
+
+    assert((newSimdBaseJitType != CORINFO_TYPE_UNDEF));
+
+    simdBaseJitType = newSimdBaseJitType;
+    simdSize        = newSimdSize;
+}
+
+//-----------------------------------------------------------------------------
+// LclMasksCheckVisitor: Find all lcl var definitions and uses. For each one, update the weighting.
+//
+class LclMasksCheckVisitor final : public GenTreeVisitor<LclMasksCheckVisitor>
+{
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    LclMasksCheckVisitor(Compiler* compiler, weight_t bbWeight, LclMasksWeightTable* weightsTable)
+        : GenTreeVisitor<LclMasksCheckVisitor>(compiler)
+        , bbWeight(bbWeight)
+        , weightsTable(weightsTable)
+    {
+    }
+
+    Compiler::fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTreeHWIntrinsic* convertOp = nullptr;
+
+        bool isLocalStore  = false;
+        bool isLocalUse    = false;
+        bool hasConversion = false;
+
+        switch ((*use)->OperGet())
+        {
+            case GT_STORE_LCL_VAR:
+                isLocalStore = true;
+
+                // Look for:
+                //      use:STORE_LCL_VAR(ConvertMaskToVector(x))
+
+                if ((*use)->AsLclVar()->Data()->OperIsConvertMaskToVector())
+                {
+                    convertOp     = (*use)->AsLclVar()->Data()->AsHWIntrinsic();
+                    hasConversion = true;
+                }
+                break;
+
+            case GT_LCL_VAR:
+                isLocalUse = true;
+
+                // Look for:
+                //      user:ConvertVectorToMask(use:LCL_VAR(x)))
+
+                if (user->OperIsConvertVectorToMask())
+                {
+                    convertOp     = user->AsHWIntrinsic();
+                    hasConversion = true;
+                }
+                break;
+
+            default:
+                break;
+        }
+
+        if (isLocalStore || isLocalUse)
+        {
+            GenTreeLclVarCommon* lclOp = (*use)->AsLclVarCommon();
+
+            // Get the existing weighting (if any).
+            LclMasksWeight weight;
+            weightsTable->Lookup(lclOp->GetLclNum(), &weight);
+
+            // Update the weights.
+            JITDUMP("Local %s V%02d at [%06u] has %s conversion. ", isLocalStore ? "store" : "var", lclOp->GetLclNum(),
+                    m_compiler->dspTreeID(lclOp), hasConversion ? "mask" : "no");
+            weight.UpdateWeight(isLocalStore, hasConversion, bbWeight);
+
+            // Cache the simd type data of the conversion.
+            if (hasConversion)
+            {
+                assert(convertOp != nullptr);
+                weight.CacheSimdTypes(convertOp);
```

Comment:
I think we have to give up here if there is a type mismatch. As it is I think you can come up with test cases that the transformation will break because of a wrongly inserted conversion. You may need `Unsafe.BitCast` for that.